onnxruntime/contrib_ops/cuda/bert/attention_past.cu

```diff
@@ -58,11 +101,23 @@ bool LaunchConcatPastToPresent(cudaStream_t stream,
                                float* present) {
   const dim3 grid(all_sequence_length, batch_size, 2);
   if (0 == (head_size & 1)) {
-    const dim3 block(head_size / 2, num_heads, 1);
-    ConcatPastToPresent<float2><<<grid, block, 0, stream>>>(sequence_length, reinterpret_cast<const float2*>(past), reinterpret_cast<const float2*>(k_v), reinterpret_cast<float2*>(present));
+    const int H = head_size / 2;
+    if (H * num_heads <= 1024) {
+      const dim3 block(H, num_heads, 1);
+      ConcatPastToPresent<float2><<<grid, block, 0, stream>>>(sequence_length, reinterpret_cast<const float2*>(past), reinterpret_cast<const float2*>(k_v), reinterpret_cast<float2*>(present));
+    } else {
+      const dim3 block(1024 / num_heads, num_heads, 1);
+      ConcatPastToPresentLarge<float2><<<grid, block, 0, stream>>>(sequence_length, H, reinterpret_cast<const float2*>(past), reinterpret_cast<const float2*>(k_v), reinterpret_cast<float2*>(present));
+    }
   } else {
-    const dim3 block(head_size, num_heads, 1);
-    ConcatPastToPresent<float><<<grid, block, 0, stream>>>(sequence_length, past, k_v, present);
+    if (head_size * num_heads <= 1024) {
```

Comment:
It is better that we replace 1024 by cudaDeviceProp::maxThreadsPerBlock. It might not be 1024 for some device. The cudaDeviceProp can be passed from QkvToContext to this function.